drivers/usb/device/usb_dc_nrfx.c

```diff
@@ -533,31 +533,49 @@ void usb_dc_nrfx_power_event_callback(nrf_power_event_t event)
  */
 static int hf_clock_enable(bool on, bool blocking)
 {
-	int ret = -ENODEV;
+	int ret = -EAGAIN;
+	/* Arbitrary number of attempts when  -EAGAIN has been returned. */
+	int attempts = 3;
 	struct device *clock;
 	static bool clock_requested;
 
 	clock = device_get_binding(DT_NORDIC_NRF_CLOCK_0_LABEL "_16M");
 	if (!clock) {
 		LOG_ERR("NRF HF Clock device not found!");
-		return ret;
+		return -ENODEV;
 	}
 
-	if (on) {
-		if (clock_requested) {
-			/* Do not request HFCLK multiple times. */
-			return 0;
+	do {
+		if (on) {
+			if (clock_requested) {
+				/* Do not request HFCLK multiple times. */
+				return 0;
+			}
+			ret = clock_control_on(clock, (void *)blocking);
+		} else {
+			if (!clock_requested) {
+				/* Cancel the operation if clock has not
+				 * been requested by this driver before.
+				 */
+				return 0;
+			}
+			ret = clock_control_off(clock, (void *)blocking);
+			if (ret == -EBUSY) {
+				/* This is an expected behaviour.
+				 * -EBUSY means that some other module has also
+				 * requested the clock to run.
+				 */
+				ret = 0;
+			}
 		}
-		ret = clock_control_on(clock, (void *)blocking);
-	} else {
-		if (!clock_requested) {
-			/* Cancel the operation if clock has not
-			 * been requested by this driver before.
+		--attempts;
+		if (ret == -EAGAIN) {
```

Comment:
I am not sure if there is a point in doing that. If thread that we wait for is of lower priority it will never run so the only case this code will cover is when two threads are of the same priority. Since it is not the most generic case I guess code could just return `-EAGAIN`.

Reply:
Also, can't this function be called from ISR? If so I don't think we should be calling `k_yield()` from it.

Reply:
I checked that this is called from places where mutex is taken so API wise nothing changes.

Reply:
Looks like it's more complicated issue. Returning -EAGAIN to application (usb_disable() -> usb_dc_detach() -> hf_clock_enable(false, false) because of clock does not make sense to me.

IMO the clock driver should be heavily reworked. It has some pseudo-mutex (m16src_grd) and it's used by BT stuff which works out of OS context. If it's a OS driver using OS API then the is should use OS primitives (mutex) and deal with out-of-context stuff and dirty hacks internally. Btw. -EAGAIN is not event handled in BT - ll_sw/ctrl.c: mayfly_xtal_start().

It's out of context if this PR, so I'll leave is as it is (without -EAGAIN handling) until the clock management driver is dealt with.